ocaml/xapi-guard/test/cache_test.ml

```diff
@@ -0,0 +1,139 @@
+let ( let@ ) f x = f x
+
+let ( let* ) = Lwt.bind
+
+module Tpm = Xapi_guard.Types.Tpm
+
+module TPMs = struct
+  let request_persist uuid write =
+    let __FUN = __FUNCTION__ in
+
+    let key = Tpm.deserialize_key (Random.int 3) in
+
+    let time = Mtime_clock.now () in
+    let contents =
+      Printf.sprintf "contents %s" (Mtime.to_uint64_ns time |> Int64.to_string)
+    in
+    let* () =
+      Logs_lwt.app (fun m ->
+          m "%s: New request! %a/%i/%a" __FUN Uuidm.pp uuid
+            Tpm.(serialize_key key)
+            Mtime.pp time
+      )
+    in
+    write (uuid, time, key) contents
```

Comment:
The mtime here is host and boot specific, i.e. after a host reboot you may get something that is now smaller.
If we are using the filesystem then we could use the timestamp from the file's `Unix.stat` perhaps (although we should whether we are getting nano-second resolution timestamps, on older OCaml versions or systems the timestamp resolution may be truncated --- don't remember exactly perhaps to multiples of seconds?)

If we do store an mtime it'd also be useful to store a unique boot id/ boot counter.

Reply:
Or if we can always guarantee that on startup we'd first flush the disk queue and only then allow new VMs to start, but we may not able to give that guarantee (e.g. if we have a huge backlog to replay).

Reply:
> The mtime here is host and boot specific, i.e. after a host reboot you may get something that is now smaller. 

This is the reason that the cache items are re-keyd with a fresh timestamp before the writing thread is started up, so it can be ensured newly produced requests do not get shadowed by contents from previous boots existing in the cache. See [`379e2c5` (#5460)](https://github.com/xapi-project/xen-api/pull/5460/commits/379e2c57fbd74f88e323032bde295bbf69d7316b)